doc/libpmemkv_json_config.3.md
---
draft: false
layout: "library"
slider_enable: true
description: ""
disclaimer: "The contents of this web site and the associated <a href=\"https://github.com/pmem\">GitHub repositories</a> are BSD-licensed open source."
aliases: ["libpmemkv_json_config.3.html"]
title: pmemkv_json_config
section: 3
secondary_title: pmemkv
---

[comment]: <> (SPDX-License-Identifier: BSD-3-Clause)
[comment]: <> (Copyright 2019-2021, Intel Corporation)

[comment]: <> (libpmemkv_json_config.3 -- man page for libpmemkv_json_config configuration API)

# Table of Contents

[NAME](#name)\
[SYNOPSIS](#synopsis)\
[DESCRIPTION](#description)\
[ERRORS](#errors)\
[EXAMPLE](#example)\
[SEE ALSO](#see-also)


# NAME #

**pmemkv_json_config** - helper configuration API for libpmemkv

# SYNOPSIS #

```c
#include <libpmemkv_json_config.h>

int pmemkv_config_from_json(pmemkv_config *config, const char *jsonconfig);
const char *pmemkv_config_from_json_errormsg(void);
```

For general description of pmemkv and available engines see **libpmemkv**(7).
For description of pmemkv core API see **libpmemkv**(3).
For description of configuration API for libpmemkv see **libpmemkv_config**(3).

# DESCRIPTION #

pmemkv_json_config is a helper library that provides two functions:

`int pmemkv_config_from_json(pmemkv_config *config, const char *jsonconfig);`

:	Parses JSON string and puts all items found in JSON into `config`. Allowed types
	in JSON strings and their corresponding types in pmemkv_config are:
	+ **number** -- int64 or uint64
	+ **string** -- const char *
	+ **object** -- (another JSON string) -> pointer to pmemkv_config (can be obtained using pmemkv_config_get_object)
	+ **True**, **False** -- int64

`const char *pmemkv_config_from_json_errormsg(void);`

:	Returns a human readable string describing the last error.

The `pmemkv_config_from_json` function depends on RapidJSON library
what is the direct cause of the creation of this small library.

The building of this library is enabled by default.
It can be disabled by setting the **BUILD_JSON_CONFIG** CMake variable to OFF:

```sh
cmake .. -DBUILD_JSON_CONFIG=OFF
```

## ERRORS ##

The `pmemkv_config_from_json()` function returns status. Possible return values are:

+ **PMEMKV_STATUS_OK** -- no error
+ **PMEMKV_STATUS_UNKNOWN_ERROR** -- unknown error
+ **PMEMKV_STATUS_CONFIG_PARSING_ERROR** -- parsing config data failed

# EXAMPLE #

An example using this API can be found in **libpmemkv_config**(3).

# SEE ALSO #

**libpmemkv**(7), **libpmemkv**(3), **libpmemkv_config**(3) and **<https://pmem.io>**


ChangeLog
Fri Sep 02 2022 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.4.1

	This is the last bugfix release for pmemkv 1.4 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- Fix get_all return value in robinhood engine
	- Fix iterator return value for zeroed keys

Mon Aug 08 2022 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.3.1

	This is the last bugfix release for pmemkv 1.3 version.
	Maintenance of this version is no longer supported.

	It contains all bugfixes and updates from previous patch releases
	(1.2.1, 1.1.1, 1.0.3).

Tue Jul 27 2021 Igor Chorążewicz <igor.chorazewicz@intel.com>

	* Version 1.5.0

	New features:
	- Introduce (optional) caching layer in radix engine. It can be enabled
		via config parameters
	- Add new config's flag - 'create_if_missing' to address the missing functionality.
		If set, pmemkv tries to open the pool and creates it if missing.
		It's mutually exclusive with 'create_or_error_if_exists' flag

	Bug fixes:
	- Return correct status for wrong parameters in pmemkv_config_from_json
		(instead of failing on assert)
	- Fix iterator::key() implementation to return the whole key, not only
		the part up to the first null-terminator
	- Return correct status from memkind-based engines on open failure

	Other changes:
	- Add alias 'create_or_error_if_exists' for 'force_create' config's flag and
		deprecate the latter one. The old flag can still be used and it behaves
		with no difference. It's discouraged from using, though
	- Add dram_vcmap engine based on tbb::concurrent_hash_map for benchmarking
		purposes
	- Switch default CMake's build type from "Debug" to "RelWithDebInfo"
		when compiling from sources - it's relevant for a GitHub users
		(who may e.g. run some benchmarks)

Tue Jul 06 2021 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.2.1

	This is the last bugfix release for pmemkv 1.2 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- fix of missing checks in move assigment operators
	- stree: fix operator== by adding missing const specifier
	- stree: fix missing check in node_iterator operator=
	- json_config: return error code for wrong parameters, instead of asserting them
	- json config: extend error messages

Mon Jun 28 2021 Łukasz Stolarczuk <lukasz.stolarczuk@intel.com>

	* Version 1.1.1

	This is the last bugfix release for pmemkv 1.1 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- json_config: return error status for improper parameters instead of
		asserting nullptr (and crashing)
	- fix error messages for NOT_FOUND and STOPPED_BY_CB statuses (which are not errors)
		by omitting setting (error) message from the previous status
	- vcmap: fix put() implementation (to be thread-safe)
		and decrease the number of temporary string allocations
	- vcmap: fix value_type for concurrent_hash_map
	- vcmap: fix to work with the latest TBB (2021.1) #894
	- stree: fix operator==
	- clarify and extend documentation

Mon Feb 15 2021 Igor Chorążewicz <igor.chorazewicz@intel.com>

	* Version 1.4

	Features:
	- new experimental engine - robinhood (concurrent, unsorted,
		persistent map)
	- experimental iterators API which allows to directly modify values
		(or parts of values)
	- experimental transaction API which allows grouping `put` and `remove`
		operations into a single atomic action

	Other changes:
	- optimized put() in radix engine

	Major fixes:
	- calling pmemkv_errormsg() now returns empty message for NOT_FOUND and
		STOPPED_BY_CB (instead of a possible previous, unrelated error)
	- fix vcmap OOM error on put (github issue #798)

	Known issues:
	- vcmap (github issue #894)

Tue Oct 06 2020 Szymon Romik <szymon.romik@intel.com>

	* Version 1.0.3

	This is the last bugfix release for pmemkv 1.0 version.
	Maintenance of this version is no longer supported.

	Notable changes:
	- VSmap::put method refactor (to use emplace)
	- add errormsg() to db class

Fri Oct 02 2020 Szymon Romik <szymon.romik@intel.com>

	* Version 1.3

	This release introduces a new experimental engine - radix (single-threaded
	sorted map, backed by libpmemobj-cpp's radix_tree container).

	We have also extended configuration class API and redesigned optimized
	version of stree engine.

	Features:
	- radix engine (single-threaded sorted map)
	- config setters with type safety for common config fields
	- stree engine optimization (single-threaded sorted map with custom
		comparator support)

	Major fixes:
	- fixed operator== in stree engine
	- fixed missing checks in move assignment operators

	Other changes:
	- operator<< for pmem::kv::status

	Known issues:
	- vcmap (github issues #623, #798)

Fri May 29 2020 Szymon Romik <szymon.romik@intel.com>

	* Version 1.2

	This release introduces a new experimental engine - csmap (concurrent
	sorted map, backed by libpmemobj-cpp's concurrent_map).

	We have also provided support for custom comparators for sorted engines.

	Features:
	- csmap engine (concurrent sorted map)
	- support for custom comparator for sorted engines

	Optimizations:
	- vsmap::put() now uses emplace() internally

	Other changes:
	- test framework refactor (scenario-based tests, extednded tests for
		concurrentcy and sorting)

Fri Feb 07 2020 Igor Chorążewicz <igor.chorazewicz@intel.com>

	* Version 1.0.2

	This is a bugfix release for pmemkv 1.0

	This release:
	- fixes misleading information in manpages
	- fixes compilation of vsmap engine on newer compilers
	- introduces requirement on size of std::string to be less than or equal to 32

Fri Jan 31 2019 Szymon Romik <szymon.romik@intel.com>

	* Version 1.1

	This release introduces a defragmentation feature and cmap engine
	optimizations.

	Features:
	- API extension for defragmentation (currently supported only by cmap engine)

	Optimizations:
	- faster restart time of cmap engine
	- faster put method execution for cmap engine

Mon Oct 28 2019 Szymon Romik <szymon.romik@intel.com>


src/libpmemkv_json_config.cc
// SPDX-License-Identifier: BSD-3-Clause
/* Copyright 2019-2021, Intel Corporation */

#include "libpmemkv_json_config.h"
#include "out.h"

#include <rapidjson/document.h>
#include <rapidjson/istreamwrapper.h>
#include <rapidjson/ostreamwrapper.h>
#include <rapidjson/writer.h>

extern "C" {

int pmemkv_config_from_json(pmemkv_config *config, const char *json)
{
	rapidjson::Document doc;
	rapidjson::Value::ConstMemberIterator itr;

	try {
		if (!config) {
			throw std::runtime_error("Config has to be specified");
		}
		if (!json) {
			throw std::runtime_error(
				"Configuration json has to be specified");
		}
		if (doc.Parse(json).HasParseError()) {
			throw std::runtime_error("Config parsing failed");
		}
		for (itr = doc.MemberBegin(); itr != doc.MemberEnd(); ++itr) {
			if (itr->value.IsString()) {
				auto value = itr->value.GetString();

				auto status = pmemkv_config_put_string(
					config, itr->name.GetString(), value);
				if (status != PMEMKV_STATUS_OK)
					throw std::runtime_error(
						"Inserting string to the config failed");
			} else if (itr->value.IsInt64()) {
				auto value = itr->value.GetInt64();

				auto status = pmemkv_config_put_int64(
					config, itr->name.GetString(), value);
				if (status != PMEMKV_STATUS_OK)
					throw std::runtime_error(
						"Inserting int to the config failed");
			} else if (itr->value.IsTrue() || itr->value.IsFalse()) {
				auto value = itr->value.GetBool();

				auto status = pmemkv_config_put_int64(
					config, itr->name.GetString(), value);
				if (status != PMEMKV_STATUS_OK)
					throw std::runtime_error(
						"Inserting bool to the config failed");
			} else if (itr->value.IsObject()) {
				rapidjson::StringBuffer sb;
				rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
				itr->value.Accept(writer);

				auto sub_cfg = pmemkv_config_new();

				if (sub_cfg == nullptr) {
					ERR() << "Cannot allocate subconfig";
					return PMEMKV_STATUS_OUT_OF_MEMORY;
				}

				auto status =
					pmemkv_config_from_json(sub_cfg, sb.GetString());
				if (status != PMEMKV_STATUS_OK) {
					pmemkv_config_delete(sub_cfg);
					throw std::runtime_error(
						"Cannot parse subconfig");
				}

				status = pmemkv_config_put_object(
					config, itr->name.GetString(), sub_cfg,
					(void (*)(void *)) & pmemkv_config_delete);
				if (status != PMEMKV_STATUS_OK)
					throw std::runtime_error(
						"Inserting a new entry to the config failed");
			} else {
				static std::string kTypeNames[] = {
					"Null",	 "False",  "True",  "Object",
					"Array", "String", "Number"};

				throw std::runtime_error(
					"Unsupported data type in JSON string: " +
					kTypeNames[itr->value.GetType()]);
			}
		}
	} catch (const std::exception &exc) {
		ERR() << exc.what();
		return PMEMKV_STATUS_CONFIG_PARSING_ERROR;
	} catch (...) {
		ERR() << "Unspecified failure";
		return PMEMKV_STATUS_CONFIG_PARSING_ERROR;
	}

	return PMEMKV_STATUS_OK;
}

const char *pmemkv_config_from_json_errormsg(void)
{
	return out_get_errormsg();
}

} /* extern "C" */
